### PR TITLE
[HIP] Add new eigensolvers from rocSOLVER

### PR DESCRIPTION
**Pull request by leofang**
~~The batched version has a segfault pending investigation (https://github.com/ROCmSoftwarePlatform/rocSOLVER/issues/295); the non-batched version passes tests.~~

***UPDATE***: Please see https://github.com/cupy/cupy/pull/5555#issuecomment-886138072 for a question.

### CODE REVIEW

**Comment by leofang** [comment]
The complex solver seems to be very strict about the input being a valid Hermitian matrix, which is a bit odd since we should be able to specify either the upper or lower part of the matrix through `uplo`. I'll investigate further but this can be merged for now.

**Comment by leofang** [comment]
The tests are kinda overhauled:
- The `argsort` part is unnecessary because everyone sorts the eigenvalues in the ascending order
- Batched capability (which already exists in cuSOLVER) is tested 

I am not sure if this part should be backported or not.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
> The complex solver seems to be very strict about the input being a valid Hermitian matrix

This was due to my silliness when writing the test code and is fixed in 78ce4b6. Everything should work fine now.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
This is ready.

**Review by toslunar — COMMENTED** [review]
I reviewed the tests.  The tests look almost good.

How about refactoring to `A = get_hermitian(xp, a, self.UPLO)` by defining
```python
def get_hermitian(xp, a, UPLO):
    assert UPLO in 'UL'
    if UPLO == 'U':
        tri = xp.triu(a, 1)
    else:
        tri = xp.tril(a, -1)
    h = tri + tri.swapaxes(-2, -1).conj()
    n = a.shape[-1]
    diag = xp.s_[..., xp.arange(n), xp.arange(n)]
    h[diag] = a[diag].real
    return h
```
(`.real` for the diagonal is required for general test inputs, e.g. `np.linalg.eigh([[2+3j]])` => `(array([2.]), array([[1.+0.j]]))`.)

(To test correctness of the eigensolvers, orthogonality/unitarity of `v` could be checked in addition to the reconstruction test.  But we are testing correctness of the wrappers.)

**Comment by leofang** [comment]
Thanks, @toslunar. Will make the change shortly. `xp.s_` is magic trick I was looking for. btw we don't need `.real`, as adding `a` with its complex conjugate `a.conj()` will guarantee the diagonal is real.

File: tests/cupy_tests/linalg_tests/test_eigenvalue.py [hunk 1]
```diff
@@ -4,30 +4,59 @@
...
 
 
+def _get_hermitian(xp, a, UPLO):
+    assert UPLO in 'UL'
+    A = xp.triu(a) if UPLO == 'U' else xp.tril(a)
+    A = A + A.swapaxes(-2, -1).conj()
+    n = a.shape[-1]
+    # Note: there is no "cupy.s_()", but we're just constructing slice objects
+    # here, so it's fine to call "numpy.s_()".
+    diag = numpy.s_[..., xp.arange(n), xp.arange(n)]
```
Comment (leofang) [inline tests/cupy_tests/linalg_tests/test_eigenvalue.py:19]: See https://github.com/cupy/cupy/issues/5567.

**Comment by leofang** [comment]
Blocked by #5569.

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/cusolver.pyx [hunk 2]
```diff
@@ -65,14 +65,14 @@ _available_cuda_version = {
 }
 
 _available_hip_version = {
-    'gesvdjBatched': (309, None),  # = rocsolver_<t>gesvd_batched
-    'potrfBatched': (306, None),
     # Below are APIs supported by CUDA but not yet by HIP. We need them here
     # so that our test suite can cover both platforms.
```
Comment (toslunar) [inline cupy/cusolver.pyx:69]: Is the comment outdated?

Comment (leofang) [reply cupy/cusolver.pyx]: Good catch @toslunar! I just updated it to align with that in `cupy/cusparse.py`.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Now neither of the ROCm CI can finish....🤔

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
The ROCm 4.2 CI is back (#5636). Let's get this PR tested one more time.

Jenkins, test this please

**Review by toslunar — APPROVED** [review]
LGTM

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 3]
```diff
@@ -903,7 +904,8 @@ def _eigen_vec_transform(block_vec, xp):
 
 @testing.with_requires('scipy>=1.4')
 @testing.gpu
-@pytest.mark.xfail(runtime.is_hip, reason='syevd not working')
+@pytest.mark.xfail(runtime.is_hip and driver.get_build_version() < 402,
```
Comment (kmaehashi) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:907]:
```suggestion
@pytest.mark.skipif(runtime.is_hip and driver.get_build_version() < 402,
```

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 4]
```diff
@@ -1200,10 +1202,18 @@ def test_maxit_None(self):
...
     'sparse_format': ['coo', 'csr', 'csc']
 }))
-@pytest.mark.xfail(runtime.is_hip, reason='either spMM or syevd not working')
 # test class for testing against diagonal matrices overall various data types
 class TestLOBPCGForDiagInput(unittest.TestCase):
 
+    def setUp(self):
+        if runtime.is_hip:
+            if driver.get_build_version() < 402:
+                pytest.xfail('syevj not available')
```
Comment (kmaehashi) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:1211]:
```suggestion
                pytest.skip('syevj not available')
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]
LGTM except for minor tweaks.

I think XFAIL should only be used for known issues of the platform (CUDA/ROCm). So that we can notice that the issue is fixed in the later versions of the platform by running the test suite (#5407).

**Comment by leofang** [comment]
Good catch @kmaehashi! Yes they should be skipped, not xfail'd. I probably did a simple text substitution without thinking 😅

**Review by kmaehashi — APPROVED** [review]
Thanks, LGTM!

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
Oh @kmaehashi I forgot we still have this question: https://github.com/cupy/cupy/pull/5555#issuecomment-886138072.

**Comment by kmaehashi** [comment]
Looks good to backport `tests/cupy_tests/linalg_tests/test_eigenvalue.py`!

**Comment by leofang** [comment]
> Looks good to backport `tests/cupy_tests/linalg_tests/test_eigenvalue.py`!

Sounds good, will send a PR later.

**Pull request merged**